### ISSUE TITLE
Questions about the TypeScript ecosystem

### ISSUE THREAD

**Issue by blakeembrey** [teppeis/typescript-simple#20]
Since I got an answer so quickly the other day from @basarat and @teppeis I figured I'd ask some more high level questions (not directly related to the repo, but they are things that are stuck in my head). 

How do you see this module being used? I would love to spin this into different projects (Webpack, node `require`, etc) since I've had issues trying every other variation of them (they should be working together, but that's a separate issue), but I wanted to hear what you thought about it. Do you see it used like this? If so, how does validation happen? Since I run into compilation issues with `/// <reference ...>` before, would this now work with validations or do you expect it to be used without validations? In which case, do you see node and webpack using the non-semantic mode and a separate process checking for type errors? Is there a way to run the typescript compiler without an output?

Secondly, how is the node ecosystem right now? I've tried publishing and working with some node modules but I've run into issues when publishing with the `d.ts` and reusing them. On top of that, `tsd` doesn't provide enough of a framework to get going (it's a cool project, but the centralised repository can't work forever) and it seems to be dead. `tsd link` is what I like, but it doesn't handle multiple dependencies (E.g. I depend on React in my module but the person requiring my module also has a React definition. Now there's a conflict with TypeScript modules.) Do you think a new TypeScript definition manager should be developed to handle the local case foremost? Does anyone know the status of `tsd` too (I shot off some emails to him, but is everything ok)?

Anyway, any answers would truly help and I want to help bring TypeScript to be a first-class citizen everywhere. The experience I've had has been quite underwelming (I think I've tried half the modules doing each thing out there and keep resorting back to the TypeScript compiler and a `.tmp` directory hack or something).

**Comment by basarat** [comment]
Okay first off ... the TS code sharing story is not good. I am working on fixing it slowly and plan to have a consistent story by the end of the year. 

> Webpack: 

I've had success with : https://github.com/andreypopp/typescript-loader 

> node require

Once TS1.5 lands it should *just work* (i.e `/// <ref` will work too). Everything else is still possible same as atom's : https://github.com/atom/atom/pull/5898/files#diff-1bb8c8dc24534560f9903363ebb9a332R78 

>  In which case, do you see node and webpack using the non-semantic mode and a separate process checking for type errors

Yes a seperate grunt build / IDE tooling (shameless plugs : https://github.com/TypeStrong/grunt-ts / https://github.com/TypeStrong/atom-typescript) 

> Secondly, how is the node ecosystem right now

Not *trivial* at the moment. 

If its JS you need to do `tsd install's`. 
If its TS you are sharing means you need to use stuff like @csnover's https://github.com/SitePen/dts-generator 
If its TS you are using you can hope they have generated a `.d.ts` file with their project. I would push for the `tsd link` spec of specifying in the `typescript` property but its not been adapted by the TS team. Also TS `require` does not resolve ... but its on the cards. TS team have a full plate at the moment but one thing on their plate is SystemJS (needed by angular team) so I am hoping a good story comes out of that. 

> Does anyone know the status of tsd too

https://github.com/Bartvds been a bit dark. But before he went offline I did see some activity on github so I am assuming *he* is okay. We can get you commit access to tsd if you want to take it over (as I am on definitely typed : https://github.com/DefinitelyTyped/)

**Comment by blakeembrey** [comment]
> I am working on fixing it slowly and plan to have a consistent story by the end of the year.

That's awesome! I've seen you about everywhere and if there's anything you need a hand with I'd love to contribute.

> I've had success with : https://github.com/andreypopp/typescript-loader

That's great to hear! I'm trying to use TypeScript 1.5 on a new project now though. It's not super important, but I like to play with the newer projects in my personal projects.

> Once TS1.5 lands it should just work (i.e /// <ref will work too).

That's awesome! What are we talking about for "just work". Like setting up an extension for node and requiring transparently will be simpler and actually work now? Does this include type information? If there's any resources I should check out for this stuff, I'd love to check it out - things seem to be a little everywhere at the moment.

By the way, `atom-typescript` is realy impressive! It's a bummer Atom is still so slow and a little buggy.

I've been using `tsd` a bit and it's a great stopgap, but I find it falls well short. I think I'm split 50/50 on projects between custom typings and including existing typings (E.g. working with the `atom` typing I've written maybe 100 lines to extend it with more of the API). Contributing these back is tricky too, having a centralised location is cool but it looks like it's struggling under the load (or will if more people start using Typescript).

I really like the idea behind `tsd link` and would love to make that the first class citizen. I think that's the ideal world to move toward - keeping type information with each project (easier to maintain than 1 repo with hundreds of issues/PRs). Even if it's not a TypeScript project.

I've looked at https://github.com/SitePen/dts-generator before but not really tried it. How has your milage been? Why can't you just use the TypeScript output directly from the `declaration` option? I tried it and ended up hacking(?) a solution for external modules, but I haven't tested on a large output yet (https://github.com/blakeembrey/react-free-style/blob/master/react-free-style.d.ts).

> We can get you commit access to tsd if you want to take it over (as I am on definitely typed : https://github.com/DefinitelyTyped/)

I would love to help on `tsd`. Definitely on refining an external module story. Let me know the best place I can contribute though :smile:

**Comment by csnover** [comment]
> Why can't you just use the TypeScript output directly from the declaration option?

The biggest reason is that the TypeScript compiler has no configurable mechanism for mapping module IDs to file paths, just its own naïve path walking algorithm (which does not match how *any* JS module system works to resolve modules), so to allow people to use the standard compiler, one must generate a d.ts file with the module IDs baked in. This prepopulates the compiler with the module definitions, so bypasses the non-standard path walking, but it also sucks for a lot of reasons, including that you’re basically screwed if you have two dependencies that use two different incompatible versions of the same package. I’m not sure what the original design parameters of the TS compiler were to end up where it is right now but it is definitely a situation needing improvement as the ecosystem grows.

**Comment by basarat** [comment]
> including that you’re basically screwed if you have two dependencies that use two different incompatible versions of the same package

@csnover  I am thinking of baking in support where we scan `node_modules` for you and add any `package.json` `definition` section and if your `definition` references any `typing` then we disregard that *if you have a similar typing (by name)*. 

Sound plausible?

**Comment by blakeembrey** [comment]
@basarat I'm just thinking out loud right now, but I think the ideal TypeScript module definition resolution would look something like this (maybe it can even be simpler?). You have a package and you install definitions from the central repository (maybe other sources in the future, that would be cool) or from your own dependencies (`package.json` or `bower.json` (not sure how bower support is right now anyway) dependencies). When you install a dependency, it should define its dependencies (which would be in the `tsd.json` or it's own `package.json`/`bower.json` depending on how we resolved it. If a namespace conflicts, we can do "bower-style" resolution that keeps the user the option to pick (most recent hash, latest version, etc.). I think that all dependencies then need to be copied flat into `typings` to properly resolve.

There would be a bunch to test and validate this, but it'd be *so* nice to have this already.

**Comment by basarat** [comment]
> not sure how bower support is right now anyway

None

**Comment by blakeembrey** [comment]
> None

Haha, ok. I haven't tried it (I don't use bower myself) but I was going off https://github.com/DefinitelyTyped/tsd#link-to-bundled-definitions

**Comment by basarat** [comment]
Cool. I've started on https://github.com/TypeStrong/atom-typescript/issues/167 and will see how it goes

**Comment by blakeembrey** [comment]
@basarat Any chance I can help out with the TSD tool? I'd love to attempt a rewrite that focuses on a flattened dependency tree to avoid conflicts and automatically support node modules and bower modules, as well as support GitHub and HTTP sources in the `tsd.json` like it does now. My only issue with doing something like this right now is that the repo is dead, so, without any confirmation to go off, I'd probably be writing my own tool.

**Comment by basarat** [comment]
@blakeembrey I've given you git push access, so you can start experimenting.

I personally don't have npm push at the moment. Will pass that on once I get that too :+1:

**Comment by blakeembrey** [comment]
@basarat Cheers! I might go through the issues during the week and work on some new ideas on the weekend. Excited to help this thing move along :smile:
